rfcs/20221209-data-cleaning.md

```diff
@@ -0,0 +1,26 @@
+# RFC: Cleaning up the upstream database
+
+One of the steps of the [data refresh process for images][data-refresh] is cleaning data that is not fit for production. This process runs weekly as Airflow DAGs, the cleaned data is only saved to the API database (DB), which is replaced when the data refresh finishes, so it needs to be cleaned every time. This cleaning step takes up to 16 hours lately. It has been stable in duration given the strategy changed to perform the validations and transformations of the data upfront, when pulling from providers rather than when copying the upstream catalog DB into the API DB, but images ingested prior to this change remained untouched.
```

Comment:
> images ingested prior to this change remained untouched.

Is this something we can account for efficiently in the cleanup DAG, to reduce the number of images that need to be cleaned?

Because many of our DAGs are not dated, and therefore re-ingest all data on each run (using the MediaStore even if they did not when originally ingested), does this mean that all records for those providers should be okay? Likewise, for our dated DAGs do we only need to be concerned about records that have not recently been reingested?

The only other exception I can think of is records which were ingested before the introduction of the MediaStores, then deleted from the provider (which is not caught by our DAGs). I also know that we have some records not associated to provider DAGs, which will definitely need to be cleaned.

Reply:
Can you confirm that the URLs are replaced with the validated URLs, and the tags are _replaced_ and not simply added to during the re-ingestion, @stacimc? Tags use a JSON Array Column that uses the following strategy when upserting:

https://github.com/WordPress/openverse-catalog/blob/5933f712d2d017eb1d952c68fffc6f1606d58eb1/openverse_catalog/dags/common/storage/columns.py#L57-L65

I can't really understand what exactly is happening here :)

Reply:
URLs (and most other data) are overwritten during ingestion, but you're right @obulat that tags are only added to. Confirmed locally:

* Ingest a record A with tag names "foo" and "bar"
* The external provider updates the record to delete/update tags, so it now has only tag name "FOO"
* Re-ingest record A
* The record in the Catalog will now contain tag names "foo", "bar", and "FOO"

I do think that feels like an additional issue -- maybe one that should be handled here? @krysal what do you think? For non-dated DAGs, if we end up changing that upsert strategy we'd either need to (a) do something like the data cleanup DAG again, or (b) commit to letting the reingestion workflows slowly update them.

That being said, as far as the steps in the data cleanup go I _think_ the point stands -- for non-dated DAGs, they should be updated during regular ingestion _once the additional steps are added to the MediaStore_. Does that make sense or do you see other edge cases?

Still the only exception I see is dead records that have been deleted from the original provider but are still in our catalog. It would be nice to get a better picture of how many of those we have.